### ISSUE TITLE
TypeError in RNNTJoint

### ISSUE THREAD

**Issue by kokamido** [NVIDIA/NeMo#6328]
**Describe the bug**
Hi. In my scenario `RNNTJoint.forward` produces TypeError and error message is 
```
narrow() received an invalid combination of arguments - got (length=Tensor, start=int, dim=int, ), but expected one of:
 * (int dim, Tensor start, int length)
      didn't match because some of the arguments have invalid types: (dim=int, start=int, length=Tensor, )
 * (int dim, int start, int length)
      didn't match because some of the arguments have invalid types: (dim=int, start=int, length=Tensor )
```

**Steps/Code to reproduce bug**
I faced this issue with nemo-toolkit==1.12 but it looks like problem still exists in the version 1.16. 
The nature of the error itself can be shown using the following code:
```python
import torch
a = torch.Tensor([0,1,1])
max_a = a.max()
a.narrow(dim=0, start=0, length = max_a)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-1104b945c0b8> in <cell line: 4>()
      2 a = torch.Tensor([0,1,1])
      3 max_a = a.max()
----> 4 a.narrow(dim=0, start=0, length = max_a)

TypeError: narrow() received an invalid combination of arguments - got (length=Tensor, start=int, dim=int, ), but expected one of:
 * (int dim, Tensor start, int length)
      didn't match because some of the arguments have invalid types: (dim=int, !start=int!, !length=Tensor!, )
 * (int dim, int start, int length)
      didn't match because some of the arguments have invalid types: (dim=int, start=int, !length=Tensor!, )
```

And there is the similar code in the [nemo 1.12](https://github.com/NVIDIA/NeMo/blob/dd9a30f55b7ec308cc3b55751e8ae5aa40f36a62/nemo/collections/asr/modules/rnnt.py#L1311
) (lines 1297 and 1311) as well as in the [nemo 1.16](https://github.com/NVIDIA/NeMo/blob/163111870e78a2a7224953a89b6af5ac71cb8ed9/nemo/collections/asr/modules/rnnt.py#L1315) (lines 1301 and 1315)

In order to reproduce this behavior you have to make the condition `sub_dec.shape[1] != max_sub_transcript_length + 1`  true. In my scenario, this is true because I use several hacks around my own custom dataset and dataloader that applies padding to the data if "time" dimension of tensor  is not a multiple of 128. But I'm quite sure there is a cleaner way to fulfill the condition, and it looks like the code will throw this error every time the condition is met.

Replacement `max_sub_transcript_length = sub_transcript_lens.max()` -> `max_sub_transcript_length = int(sub_transcript_lens.max().item())` applied [here](https://github.com/NVIDIA/NeMo/blob/dd9a30f55b7ec308cc3b55751e8ae5aa40f36a62/nemo/collections/asr/modules/rnnt.py#L1297) fixes the issue for me.

**Expected behavior**

The line `sub_dec = sub_dec.narrow(dim=1, start=0, length=max_sub_transcript_length + 1)` can be executed without rising TypeError

**Environment overview (please complete the following information)**

 - Environment location: Docker, image is based on cuda:11.3.1-cudnn8-devel-ubuntu20.04
 - Method of NeMo install: `pip install nemo-toolkit['all']==1.12.0`

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- Ubuntu20.04
- PyTorch version 1.10.1 (but the cause of the problem persists in 1.13.1 )
- Python version 3.9.7

**Additional context**
I have an additional question about the code I mentioned in this issue. There is a couple of comments that says:
```
# Reduce decoder length to preserve computation
# Decoder: [sub-batch, U, D] -> [sub-batch, U', D]; U' < U
    if sub_dec.shape[1] != max_sub_transcript_length + 1:
        sub_dec = sub_dec.narrow(dim=1, start=0, length=max_sub_transcript_length + 1)
```
I understand that we can remove the padding here in order to reduce memory consumption and computation time of the joint layer but we (probably) still have a lot of padding in the decoder output if our batch contains transcriptions of different lengths. As afs as I know current implementation of RNNT will process this padding in the joint layer and there is no way to reduce related computation and memory costs. Is it correct?

**Comment by titu1994** [comment]
This is an interesting find, I'll send a patch tomorrow. Thanks for finding it ! One issue is .item() will constantly perform GPU - CPU transfer so the fix isn't so clean but I suppose it will activate only if it enters the if branch anyway so the cost will be small overall. 

As to your final note, yes, if the batch has extremely different seq lengths then there will be padding and joint cannot avoid computation. There are ways of pre sorting the seq length, split and do fused joint forward, and then remap the IDs to original batch, but it turns out the speedup is quite negligible and memory savings are trivial after doing all that complex work because you usually use a fuser batch size << acoustic batch size, the cost of doing adaptive batch remapping is more significant than just calculating the excess padding on the joint (plus the memory usage is only trivially larger).

**Comment by kokamido** [comment]
I did some research on this problem and found out that the true cause of the error is dtype of [sub_transcript_lens](https://github.com/NVIDIA/NeMo/blob/dd9a30f55b7ec308cc3b55751e8ae5aa40f36a62/nemo/collections/asr/modules/rnnt.py#L1311) tensor. As I mentioned before, I've used a few hacks around dataloader, so sub_transcript_lens has `dtype==torch.float32` because of my hacks, while NeMo data pipeline handles length using int dtype. This is reasonable because the NeMo conformer implementation uses absolute length values.
Error message from .narrow method is quite confusing because it says that the cause of the problem is the fact that the type of the "length" argument is tensor but it actually works with tensors if dtype of tensor is int (see picture below).

Thus, it seems that the applied fix did not break anything, and it is suitable for my case (but my case is quite strange 😅 ).

![image](https://user-images.githubusercontent.com/7046074/229070481-692984ce-7fee-4c12-a78e-c24a481fdaf4.png)

**Comment by titu1994** [comment]
Could I ask what is your use case? Float len will not be accepted by the RNNT loss

**Comment by kokamido** [comment]
As I mentioned earlier, my setup was quite unusual. I used custom data loader and I have applied several hacks to inject it. My dataloader  uses relative audio length, but NeMO RNNT uses absolute length, so I calculate absolute length from relative values in the dataloader, but I didn't pay attention to dtype of corresponding tesor.
Incompatibility with RNNT loss was not a problem in this case because the issue I mentioned led to an error before calculating the loss.
